File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
p=[i for i in range(201)]
def f(x):
    if x==p[x]:
        return x
    p[x]=f(p[x])
    return p[x]
def u(x,y):
    x=f(x)
    y=f(y)
    if x!=y:
        p[y]=x
        return True
    return False

def solution(n, computers):
    answer = n
    for i in range(len(computers)):
        for j in range(i+1,n):
            if computers[i][j]:
                if u(i+1,j+1):
                    answer-=1
    return answer

File: 프로그래머스/lv3/43162. 네트워크/네트워크.cpp
#include <bits/stdc++.h>
using namespace std;
int parent[201];
int find(int x){
    if (x==parent[x])
        return x;
    return parent[x]=find(parent[x]);
}
int uni(int x,int y)
{
    x=find(x);
    y=find(y);
    if (x!=y)
    {
        parent[y]=x;
        return true;
    }
    return false;
}
int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    for (int i=1;i<n;i++)
        parent[i]=i;
    for (int i=0;i<n;i++)
    {
        for (int j=0;j<n;j++)
        {
            if (computers[i][j] && i!=j)
            {
                if (uni(i,j))answer+=1;
            }
        }
    }
    return n-answer;
}